native/dsteqr.go

```diff
@@ -0,0 +1,394 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import (
+	"math"
+
+	"github.com/gonum/blas"
+	"github.com/gonum/blas/blas64"
+	"github.com/gonum/lapack"
+)
+
+// Dsteqr computes the eigenvalues and optionally the eigenvectors of a symmetric
+// tridiagonal matrix using the implicit QL or QR method. The eigenvectors of a
+// full or band symmetric matrix can also be found if Dsytrd, Dsptrd, or Dsbtrd
+// have been used to reduce this matrix to tridiagonal form.
+//
+// d, on entry, contains the diagonal elements of the tridiagonal matrix. On exit,
+// d contains the eigenvalues in ascending order. d must have length n and
+// Dsteqr will panic otherwise.
+//
+// e, on entry, contains the off-diagonal elements of the tridiagonal matrix on
+// entry, and is overwritten during the call to Dsteqr. e must have length n-1 and
+// Dsteqr will panic otherwise.
+//
+// z, on entry, contains the n×n orthogonal matrix used in the reduction to
+// tridiagonal form if compz == lapack.EigBoth. On exit, if
+// compz == lapack.EigBoth, z contains the orthonormal eigenvectors of the
+// original symmetric matrix, and if compz == lapack.EigDecomp, z contains the
+// orthonormal eigenvectors of the symmetric tridiagonal matrix. z is not used
+// if compz == lapack.EigValueOnly.
+//
+// work must have length at least max(1, 2*n-2) if the eigenvectors are computed,
+// and Dsteqr will panic otherwise.
+func (impl Implementation) Dsteqr(compz lapack.EigComp, n int, d, e, z []float64, ldz int, work []float64) (ok bool) {
+	if len(d) < n {
+		panic(badD)
+	}
+	if len(e) < n-1 {
+		panic(badE)
+	}
+	if compz != lapack.EigValueOnly && compz != lapack.EigBoth && compz != lapack.EigDecomp {
+		panic(badEigComp)
+	}
+	if compz != lapack.EigValueOnly {
+		if len(work) < max(1, 2*n-2) {
+			panic(badWork)
+		}
+		checkMatrix(n, n, z, ldz)
+	}
+
+	var icompz int
+	if compz == lapack.EigDecomp {
+		icompz = 1
+	} else if compz == lapack.EigBoth {
+		icompz = 2
+	}
+
+	if n == 0 {
+		return true
+	}
+	if n == 1 {
+		if icompz == 2 {
```

Comment:
I'm pretty sure the code is correct. The reason is that if compz == 'V' then z already contains a valid diagonalization (according to any of the metrics specified). If compz == 'I', then z can contain anything and z must be set to I. Since size(z) = 1, it's just the first element.